### DISCUSSION TITLE
How are types encoded into the IR.

### DISCUSSION THREAD

**Discussion by DamianReeves** [finos/morphir#53] [Q&A] [answered]
In the **What does the data format look like?** section we have the following example:

> ```javascript
> quantity * unitPrice
> ```

In Morphir's data format this would translate into something like this:

> ```javascript
> ["Apply"
> , ["Apply"
>  , ["Reference", [["Morphir", "SDK"], ["Number"], "multiply"]]
>  , ["Variable", ["quantity"]]]
> , ["Variable", ["unit", "price"]]
> ]
> ```

Can you show an example of how the type info for this would be represented in the IR?

**Comment by AttilaMihaly** [comment, marked as answer]
Maybe it should be mentioned in that page that this is a simplified JSON just to illustrate the concept. The real JSON will be much more complex. For example, the above expression will translate to this JSON (it's too big with proper indentation so I included the compact version here): 

```json
["apply",["reference",null,[[["morphir"],["s","d","k"]],[["basics"]],["float"]],[]],["apply",["function",null,["reference",null,[[["morphir"],["s","d","k"]],[["basics"]],["float"]],[]],["reference",null,[[["morphir"],["s","d","k"]],[["basics"]],["float"]],[]]],["reference",["function",null,["reference",null,[[["morphir"],["s","d","k"]],[["basics"]],["float"]],[]],["function",null,["reference",null,[[["morphir"],["s","d","k"]],[["basics"]],["float"]],[]],["reference",null,[[["morphir"],["s","d","k"]],[["basics"]],["float"]],[]]]],[[["morphir"],["s","d","k"]],[["basics"]],["multiply"]]],["variable",["reference",null,[[["morphir"],["s","d","k"]],[["basics"]],["float"]],[]],["quantity"]]],["variable",["reference",null,[[["morphir"],["s","d","k"]],[["basics"]],["float"]],[]],["unit","price"]]]
```

You can also use [this page](https://morphir.finos.org/try-morphir) to generate the IR for any Elm models. It shows a visual version of the IR but there's a newer version that can show the JSON too (I have an [open PR](https://github.com/finos/morphir/pull/54) to publish it).

In general the best way to understand the JSON is to look at the [Elm API for the IR](https://package.elm-lang.org/packages/finos/morphir-elm/latest#morphir-ir) together with [the standard Morphir-to-JSON mapping](https://github.com/finos/morphir-elm/blob/master/docs/json-mapping.md). The JSON is a direct serialization of the Elm data structures using that standard mapping.

**Reply by DamianReeves** [reply]
I approved the above mentioned PR.

**Comment by AttilaMihaly** [comment]
To elaborate a bit more: You will notice that the IR JSON contains type information on every value node. This is because type information is used in many of our backends but it will make your life much more difficult when you are building a frontend. 

To avoid that I think we should expose the type inference tooling separately so that you can focus on creating the IR without type information in your frontend and then reuse the type inference we already have in the Elm tooling to enrich it with types. That would also be a nice way to check the validity of the IR you created.
